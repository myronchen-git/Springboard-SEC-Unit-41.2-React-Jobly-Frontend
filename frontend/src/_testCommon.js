const companies = Object.freeze([
  Object.freeze({
    handle: 'comp1',
    name: 'Company 1',
    description: 'Company 1 description.',
    numEmployees: 12,
    logoUrl: 'logo1.png',
  }),
  Object.freeze({
    handle: 'comp2',
    name: 'Company 2',
    description: 'Company 2 description.',
    numEmployees: 123,
    logoUrl: 'logo2.png',
  }),
  Object.freeze({
    handle: 'comp3',
    name: 'Company 3',
    description: 'Company 3 description.',
    numEmployees: 3,
  }),
  Object.freeze({
    handle: 'comp4',
    name: 'Company 4',
    description: 'Company 4 description.',
    logoUrl: 'logo4.png',
  }),
]);

const jobs = Object.freeze([
  Object.freeze({
    id: '1',
    title: 'Job 1',
    salary: 1,
    equity: 0.1,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
  Object.freeze({
    id: '2',
    title: 'Job 2',
    salary: 2,
    equity: 0.2,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
]);

const comp1Jobs = [];
[jobs[0], jobs[1]].forEach((job) => {
  const condensedJob = { ...job };
  delete condensedJob.companyHandle;
  delete condensedJob.companyName;
  comp1Jobs.push(Object.freeze(condensedJob));
});

const companyDetails = Object.freeze({
  handle: 'comp1',
  name: 'Company 1',
  description: 'Company 1 description.',
  numEmployees: 12,
  logoUrl: 'logo1.png',
  jobs: Object.freeze(comp1Jobs),
});

// ==================================================

export { companies, companyDetails, jobs };
