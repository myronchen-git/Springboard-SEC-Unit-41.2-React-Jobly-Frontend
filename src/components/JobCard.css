.JobCard .card-body {
  text-align: left;
}

.JobCard .card-subtitle {
  margin-bottom: 1em;
}

.JobCard__form {
  padding: 1em;
  text-align: right;
}
