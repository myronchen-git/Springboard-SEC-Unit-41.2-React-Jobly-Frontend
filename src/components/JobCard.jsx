import PropTypes from 'prop-types';
import { useContext } from 'react';

import { UserContext } from '../contexts.jsx';

// ==================================================

/**
 * Displays some info about a job.  Also allows applying to job.
 *
 * @param {Object} props - React component properties.
 * @param {Object} props.job - The job to display.
 * @param {Object} props.job.id - ID of the job.
 * @param {Object} props.job.title - Title of the job.
 * @param {Object} props.job.salary - Salary of the job.
 * @param {Object} props.job.equity - Equity of the job.
 * @param {Object} props.job.companyName - Name of the company that the job
 *   belongs to.
 * @param {Object} props.isApplied - Whether the user has already applied to the
 *   provided job.
 */
function JobCard({ job, isApplied }) {
  const { applyToJob } = useContext(UserContext);

  return (
    <article className="JobCard">
      <header className="JobCard__header">
        <h1>{job.title}</h1>
        {job.companyName && <p>{job.companyName}</p>}
      </header>
      <summary className="JobCard__info">
        <p>Salary: {job.salary}</p>
        <p>Equity: {job.equity}</p>
      </summary>
      <form className="JobCard__form">
        {isApplied ? (
          <button type="button" disabled>
            Applied
          </button>
        ) : (
          <button type="button" onClick={() => applyToJob(job.id)}>
            Apply
          </button>
        )}
      </form>
    </article>
  );
}

// ==================================================

JobCard.propTypes = {
  job: PropTypes.object.isRequired,
  isApplied: PropTypes.bool.isRequired,
};

export default JobCard;
