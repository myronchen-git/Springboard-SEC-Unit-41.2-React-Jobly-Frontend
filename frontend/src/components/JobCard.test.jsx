import { render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { jobs } from '../_testCommon.js';
import JobCard from './JobCard.jsx';

// ==================================================

vi.mock('react-router-dom');

// ==================================================

describe('JobCard', () => {
  const originalJob = jobs[0];
  let job;

  beforeEach(() => {
    job = structuredClone(jobs[0]);
  });

  it('renders.', () => {
    render(<JobCard job={job} isApplied={false} />);
  });

  it('matches snapshot.', () => {
    const { asFragment } = render(<JobCard job={job} isApplied={false} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays job info.', () => {
    // Act
    const { getByText } = render(<JobCard job={job} isApplied={false} />);

    // Assert
    expect(getByText(originalJob.title)).toBeVisible();
    expect(getByText('Salary: ' + originalJob.salary)).toBeVisible();
    expect(getByText('Equity: ' + originalJob.equity)).toBeVisible();
    expect(getByText(originalJob.companyName)).toBeVisible();
  });

  it('displays job info without salary.', () => {
    // Arrange
    delete job.salary;

    // Act
    const { getByText, queryByText } = render(
      <JobCard job={job} isApplied={false} />
    );

    // Assert
    expect(getByText(originalJob.title)).toBeVisible();
    expect(
      queryByText('Salary: ' + originalJob.salary)
    ).not.toBeInTheDocument();
    expect(getByText('Equity: ' + originalJob.equity)).toBeVisible();
    expect(getByText(originalJob.companyName)).toBeVisible();
  });

  it('displays job info without equity.', () => {
    // Arrange
    delete job.equity;

    // Act
    const { getByText, queryByText } = render(
      <JobCard job={job} isApplied={false} />
    );

    // Assert
    expect(getByText(originalJob.title)).toBeVisible();
    expect(getByText('Salary: ' + originalJob.salary)).toBeVisible();
    expect(
      queryByText('Equity: ' + originalJob.equity)
    ).not.toBeInTheDocument();
    expect(getByText(originalJob.companyName)).toBeVisible();
  });

  it.each([
    [false, 'Apply'],
    [true, 'Applied'],
  ])(
    'displays Apply button correctly when isApplied is %s.',
    (isApplied, expectedApplyBtnText) => {
      // Act
      const { getByText } = render(<JobCard job={job} isApplied={isApplied} />);

      // Assert
      const applyBtn = getByText(expectedApplyBtnText);

      expect(applyBtn).toBeVisible();

      isApplied
        ? expect(applyBtn).toBeDisabled()
        : expect(applyBtn).toBeEnabled();
    }
  );
});
