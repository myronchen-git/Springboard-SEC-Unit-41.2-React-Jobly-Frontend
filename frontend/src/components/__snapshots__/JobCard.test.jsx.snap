// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JobCard > matches snapshot. 1`] = `
<DocumentFragment>
  <article
    class="JobCard"
  >
    <header
      class="JobCard__header"
    >
      <h1>
        Job 1
      </h1>
      <p>
        Company 1
      </p>
    </header>
    <summary
      class="JobCard__info"
    >
      <p>
        Salary: 1
      </p>
      <p>
        Equity: 0.1
      </p>
    </summary>
    bound HTMLFormElement {
      "0": <button
        type="button"
      >
        Apply
      </button>,
    }
  </article>
</DocumentFragment>
`;
